app/schemas/tech.relaycorp.courier.data.database.AppDatabase/1.json

```diff
@@ -0,0 +1,94 @@
+{
+  "formatVersion": 1,
+  "database": {
+    "version": 1,
+    "identityHash": "fb3a04584d10d46fc5bb13be2ccfd61f",
+    "entities": [
+      {
+        "tableName": "cargo",
+        "createSql": "CREATE TABLE IF NOT EXISTS `${TABLE_NAME}` (`recipientAddress` TEXT NOT NULL, `messageId` TEXT NOT NULL, `creationTime` INTEGER NOT NULL, `expirationTime` INTEGER NOT NULL, `ttl` INTEGER NOT NULL, `storagePath` TEXT NOT NULL, `size` INTEGER NOT NULL, `senderAddress` TEXT NOT NULL, PRIMARY KEY(`recipientAddress`, `messageId`))",
+        "fields": [
+          {
+            "fieldPath": "recipientAddress",
+            "columnName": "recipientAddress",
+            "affinity": "TEXT",
+            "notNull": true
+          },
+          {
+            "fieldPath": "messageId",
+            "columnName": "messageId",
+            "affinity": "TEXT",
+            "notNull": true
+          },
+          {
+            "fieldPath": "creationTime",
+            "columnName": "creationTime",
+            "affinity": "INTEGER",
+            "notNull": true
+          },
+          {
+            "fieldPath": "expirationTime",
+            "columnName": "expirationTime",
+            "affinity": "INTEGER",
+            "notNull": true
+          },
+          {
+            "fieldPath": "ttl",
+            "columnName": "ttl",
+            "affinity": "INTEGER",
+            "notNull": true
+          },
```

Comment:
It feels a bit redundant to include the TTL whilst also storing the expiration time -- should we omit this one?

Reply:
I thought about that. We probably don't need the TTL there for any functionality we thought of so far. But the dates run on milliseconds and TTL on seconds. This way I'm sure to be able to recreate a Cargo exactly like it was before... Either way, I'm cool with it

Reply:
Millisecond precision isn't used in RAMF messages, so milliseconds will always be `0`.

> (...) I'm sure to be able to recreate a Cargo exactly like it was before

Do you mean a `StoredCargo` instance? Note that the courier won't be able to create cargo because it wouldn't have the private key of the gateway that created it, so the courier app should always treat it as a read-only file.

Reply:
I wasn't thinking of creating brand new Cargo, just re-creating a Cargo/Message instance based on the `StoredCargo` and the payload stored in a file, so it can be sent somewhere.

Reply:
> so it can be sent somewhere

Are you referring to sending the cargo to a gateway via gRPC, or somewhere else within the app or device?

If you're thinking of the former, the function to send cargo to a gateway has to take the original cargoes as`ByteArray`s/`FileInputStream`s anyway, not `Cargo` instances. There are two reasons for this:

- We can't (reliably) reconstruct the original `ByteArray` for a cargo given its `Cargo` instance. That instance doesn't hold enough data to fully reconstruct it, but even if it did, the serialization libraries we use in TypeScript and Kotlin may produce equivalent yet slightly different byte sequences, which will invalidate the digital signature -- So they gateway will refuse the resulting cargo.
- Even if we could reconstruct the original `ByteArray` from a `Cargo` instance on-the-fly, it'd be a lot more expensive than simply reading it from disk.

For this reason, we should only initialise `Cargo` instances when we're parsing them (via `Cargo.deserialize(byteArray)`) or generating it (and have the private key to do so).

Reply:
Got it, makes sense then, no need to recreated things. Then no reason to keep TTL, we can always calculate it on a query.